### PR TITLE
mps: add FlashAttention-2 variable-length op with GQA, sliding-window, and ALiBi

### PR DESCRIPTION
**Pull request by anagnorisis2peripeteia**
Tiled Metal FlashAttention-2 kernels for packed variable-length sequences with SIMD-group matrix multiply-accumulate (SGMMA) 8x8 tile MMA on Apple Silicon.

## Features

- Variable-length sequences (cu_seqlens) for packed batches
- GQA (grouped-query attention): kv_num_heads != q_num_heads
- Sliding window attention and ALiBi positional bias
- Head dims 32-256 (explicit instantiation, no runtime padding)
- Forward + backward (dQ, dK, dV), fp16/bf16/fp32

## Implementation

Three kernel paths, selected automatically by D and dtype:

**SGMMA kernels** (D=64/128, fp16/bf16):
- `flash_attn_varlen_fwd_sgmma`: 8x8 MMA tiles (WM=4 simdgroups, BQ=32, BK=16/32); row-major K layout; cooperative threadgroup loading via `BlockLoaderT` (synchronous; `simdgroup_async_copy` was removed in Metal 4.0 / macOS 26).
- `flash_attn_varlen_bwd_dq_sgmma`: K-inner loop, recomputes softmax from LSE, accumulates dQ via outer-product MMA.
- `flash_attn_varlen_bwd_dkdv_sgmma`: K-outer loop (one threadgroup per K-tile), iterates Q-blocks; col-major Q/dO smem, row-major K/V smem.
- `VarlenAttnParams` struct (`Attention.h`) bundles 7 scalar kernel args into a single constant-buffer binding (reduces per-dispatch buffer count from 15 to 9).

**EPL scalar kernels** (fp32 or D not in {64,128}): tiled online softmax, `simd_sum()` for warp-wide dot products; fallback for all head dims.

**Preprocess pass** (all paths): `D_vec[h * total_q + row] = rowsum(dO * O)`, one threadgroup per Q-block, used by both backward kernels.

## Benchmarks

**Machine:** Apple M4 Pro (48 GB), macOS 26.0 (Sequoia)
**Method:** `torch.utils.benchmark.Timer` + `blocked_autorange(min_run_time=2.0)`, `torch.mps.synchronize()` in stmt
**Baseline:** `F.scaled_dot_product_attention` on padded tensors (KV heads expanded for GQA, causal=True). This is the actual code path used without varlen support.

All times: **median +/- IQR in microseconds (us)**

### Forward

| Config | varlen (us) | padded SDPA (us) | speedup |
|---|---|---|---|
| B4 S=512 H32 D128 GQA4 | 2489.7 +/- 16.6 | 7958.6 +/- 18.5 | **3.20x** |
| B2 S=1024 H32 D128 GQA4 | 3843.9 +/- 25.1 | 15890.8 +/- 152.2 | **4.13x** |
| B1 S=2048 H32 D128 GQA4 | 6550.5 +/- 106.9 | 31343.5 +/- 385.5 | **4.78x** |
| B1 S=4096 H32 D128 GQA4 | 23364.4 +/- 191.9 | 128744.0 +/- 2371.8 | **5.51x** |
| B1 S=8192 H32 D128 GQA4 | 89377.0 +/- 3277.0 | 551722.6 +/- 7014.2 | **6.17x** |
| B4 S=512 H8 D64 GQA1 | 531.2 +/- 29.1 | 1826.6 +/- 84.8 | **3.44x** |

### Forward + Backward

| Config | varlen (us) | padded SDPA (us) | speedup |
|---|---|---|---|
| B4 S=512 H32 D128 GQA4 | 14737.4 +/- 197.6 | 23922.1 +/- 156.2 | **1.62x** |
| B2 S=1024 H32 D128 GQA4 | 25048.0 +/- 326.1 | 42318.7 +/- 235.2 | **1.69x** |
| B1 S=2048 H32 D128 GQA4 | 43718.2 +/- 948.9 | 79314.7 +/- 273.4 | **1.81x** |
| B1 S=4096 H32 D128 GQA4 | 158844.5 +/- 3506.5 | 311334.5 +/- 934.3 | **1.96x** |
| B1 S=8192 H32 D128 GQA4 | 642890.9 +/- 970.3 | FAILS | -- |
| B4 S=512 H8 D64 GQA1 | 1828.8 +/- 44.4 | 4659.2 +/- 60.7 | **2.55x** |

Padded backward at S=8192 fails with `MPSGraph does not support tensor dims larger than INT_MAX`; the varlen kernel handles it correctly.

Speedup scales with sequence length: at S=8192, the padded path allocates 4x the memory of packed sequences and performs O(S^2) wasted compute on padding tokens.

## Tests

`test_mps.py::TestMPS::test_flash_attention_varlen` correctness suite (44/44 pass):
- fp16/bf16, D=64/128, causal/non-causal, S=64-1024
- Variable-length batches (mixed sequence lengths)
- GQA=4 (H=32 kH=8, Llama-3 style)
- Backward dQ/dK/dV vs float32 SDPA reference

Runnable benchmark: `benchmarks/mps/bench_flash_attn_varlen.py`

cc @drisspg @malfet

### CODE REVIEW

**Comment by anagnorisis2peripeteia** [comment]
Re the FC window warning: this adds a new **private** MPS-specific op (`_scaled_dot_product_flash_attention_varlen_for_mps`), following the exact same precedent as `_scaled_dot_product_flash_attention_for_mps` already in the codebase. No pre-existing Python frontend code references this op — all Python usage is in tests added in this same PR. The op is not part of any public API surface, so the two-week FC split is not required here.

**Comment by malfet** [comment]
@anagnorisis2peripeteia can you resolve merge conflict first?

**Comment by liangel-02** [comment]
@anagnorisis2peripeteia thanks for your contribution! there's an existing varlen API here (https://github.com/pytorch/pytorch/blob/main/torch/nn/attention/varlen.py) can you add MPS as a backend for this instead of writing a separate _scaled_dot_product_flash_attention_varlen_for_mps. for testing can you extend existing tests in https://github.com/pytorch/pytorch/blob/main/test/test_varlen_attention.py

**Comment by anagnorisis2peripeteia** [comment]
Thanks for the feedback! Addressing both points:

**Merge conflict**: The conflict in Attention.metal is from the upstream refactor that moved the existing SDPA kernels into DecodeAttention.h. I'll resolve this by extracting the varlen FA kernels into a new VarlenAttention.h header and updating the metal compilation, keeping the structure consistent with the new layout.

**varlen.py integration**: Agree this is the cleaner approach. The plan:
1. Keep the underlying `_scaled_dot_product_flash_attention_varlen_for_mps` aten op as the implementation detail
2. Add MPS dispatch in `torch.nn.attention.varlen._varlen_attn` alongside the existing cuDNN/flash-attn paths  
3. Update tests to use `varlen_attn()` interface from `test_varlen_attention.py`
4. Keep the aten op private — users interact only through `varlen_attn`

The MPS path would handle the GQA, sliding-window, and ALiBi parameters that `varlen_attn` already accepts. `seqused_k` and `block_table` (paged attention) would raise `RuntimeError` on MPS since those aren't implemented yet, matching the existing cuDNN behaviour for unsupported options.

Will push the updated branch shortly.

**Comment by anagnorisis2peripeteia** [comment]
Addressed @liangel-02's feedback:

1. **Added MPS as a backend in `torch/nn/attention/varlen.py`** — both `_varlen_attn` and `_varlen_attn_out` now dispatch to `_scaled_dot_product_flash_attention_varlen_for_mps` when the input tensors are on MPS. Unsupported parameters (`seqused_k`, `block_table`, `num_splits`) raise `RuntimeError`.

2. **Extended `test/test_varlen_attention.py`** with three MPS-specific tests (decorated `@onlyMPS`):
   - `test_mps_varlen_correctness` — compares `varlen_attn` output against `F.scaled_dot_product_attention` as reference, for full attention and causal attention (float16)
   - `test_mps_varlen_backward` — verifies gradients flow through the backward pass
   - `test_mps_varlen_out` — verifies `varlen_attn_out` writes correctly into a pre-allocated buffer

   Also changed `instantiate_device_type_tests` to `allow_mps=True` so these tests are picked up by the MPS CI runner.

Also resolved the merge conflict in `Attention.metal` from the previous push (the main-branch refactor that moved the original SDPA kernels into `DecodeAttention.h`).

**Comment by anagnorisis2peripeteia** [comment]
Rebased cleanly onto current main.

Root cause of the CI failures: when we merged origin/main (which had refactored the SDPA decode kernels into `DecodeAttention.h` and added `PrefillAttention.h`), the conflict resolution accidentally dropped the non-varlen `flash_attn_fwd_*` kernel template implementations from `Attention.metal`. This caused `flash_attn_fwd_half_64` to not be found in the Metal library at runtime, failing `test_flash_bwd_fp16_causal`.

Fix: squashed our commits into a single clean commit on top of current main. `Attention.metal` now correctly contains:
- `#include <ATen/native/mps/kernels/DecodeAttention.h>` (main's decode kernels)
- `#include <ATen/native/mps/kernels/PrefillAttention.h>` (main's new prefill kernels)
- Our non-varlen `flash_attn_fwd/bwd` kernel templates + instantiations
- Our varlen `flash_attn_varlen_fwd/bwd` kernel templates + instantiations

**Review by drisspg — CHANGES_REQUESTED** [review]
Why new functions instead of new dispatch keys for existing

**Comment by anagnorisis2peripeteia** [comment]
Thanks for the review! On the question of why new private ops rather than new dispatch keys on existing ops:

The existing `_scaled_dot_product_flash_attention` (CUDA) and `_scaled_dot_product_flash_attention_varlen_forward` ops carry CUDA-specific outputs that Metal doesn't have — `philox_seed`, `philox_offset`, `rng_state`, and `return_debug_mask`. Bolt-on MPS dispatch to those ops would mean either (a) returning dummy tensors for the CUDA-specific outputs on every MPS forward pass, or (b) changing the public op signature, which risks breaking existing CUDA users downstream.

The precedent for private `_for_<backend>` ops is already in the codebase: `_scaled_dot_product_flash_attention_for_cpu` / `_scaled_dot_product_flash_attention_for_cpu_backward` follow the same pattern and are dispatched from the same `torch.nn.functional.scaled_dot_product_attention` Python entry point. We follow the same pattern here.

The Python integration (`torch/nn/attention/varlen.py`) routes to these private ops via the existing dispatch path, as `liangel-02` requested in the earlier review.

Happy to restructure if there's a preferred approach — just wanted to make the design rationale explicit.

File: aten/src/ATen/native/mps/kernels/Attention.metal [hunk 1]
```diff
@@ -7,3 +7,1059 @@
...
+    const constant uint&  gqa [[buffer(10)]],
+    const constant uint&  nh  [[buffer(11)]],
+    const constant float& sc  [[buffer(12)]],
+    const constant bool&  ic  [[buffer(13)]],
+    const constant uint4& qs  [[buffer(14)]],
+    const constant uint4& ks  [[buffer(15)]],
+    const constant uint4& vs  [[buffer(16)]],
+    const constant uint4& os  [[buffer(17)]],
+    const constant uint4& dos [[buffer(18)]],
+    const constant uint4& dqs [[buffer(19)]],
```
Comment (malfet) [inline aten/src/ATen/native/mps/kernels/Attention.metal:180]: Use header files shared between host and GPU shaders to pass structured elements. Also binding 15 buffers is somewhat expensive

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -15316,6 +15316,28 @@
...
+- func: _scaled_dot_product_flash_attention_varlen_for_mps(Tensor query, Tensor key, Tensor value, Tensor cum_seq_q, Tensor cum_seq_k, SymInt max_q, SymInt max_k, float dropout_p, bool is_causal, *, float? scale=None, SymInt? window_size_left=None, SymInt? window_size_right=None, Tensor? alibi_slopes=None) -> (Tensor output, Tensor logsumexp)
+  dispatch:
+    MPS: _scaled_dot_product_flash_attention_varlen_for_mps
+  tags: nondeterministic_seeded
+
+- func: _scaled_dot_product_flash_attention_varlen_for_mps_backward(Tensor grad_out, Tensor query, Tensor key, Tensor value, Tensor out, Tensor logsumexp, Tensor cum_seq_q, Tensor cum_seq_k, SymInt max_q, SymInt max_k, float dropout_p, bool is_causal, *, float? scale=None, SymInt? window_size_left=None, SymInt? window_size_right=None, Tensor? alibi_slopes=None) -> (Tensor grad_query, Tensor grad_key, Tensor grad_value)
+  device_check: NoCheck
+  variants: function
+  dispatch:
+    MPS: _scaled_dot_product_flash_attention_varlen_for_mps_backward
```
Comment (malfet) [inline aten/src/ATen/native/native_functions.yaml:15338]: What is this? Why new dispatch rather than new key on the existing one?

File: aten/src/ATen/native/mps/kernels/Attention.metal (hunk 1 above)
Comment (anagnorisis2peripeteia) [reply aten/src/ATen/native/mps/kernels/Attention.metal:180]: The buffer-per-scalar approach here follows `sdpa_vector` in `DecodeAttention.h` which uses the same individual-binding style. That said, `PrefillAttention.h` (added more recently in #181575) uses the `PrefillAttnParams` struct pattern you're describing — happy to refactor to that approach for consistency with the newer style. Just wanted to confirm direction before making the change.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (anagnorisis2peripeteia) [reply aten/src/ATen/native/native_functions.yaml:15338]: Same rationale as the thread with @drisspg above — the existing `_scaled_dot_product_flash_attention` CUDA op carries `philox_seed`/`philox_offset`/`rng_state`/`return_debug_mask` outputs that don't exist on Metal. Adding MPS dispatch there would mean returning dummy tensors for all of them on every forward pass. The `_scaled_dot_product_flash_attention_for_cpu` precedent follows the same private-op pattern and dispatches from the same `F.scaled_dot_product_attention` entry point.

**Comment by anagnorisis2peripeteia** [comment]
@drisspg — following up on your review. The rationale for the separate private op () was explained in the comment above: the CUDA varlen forward op carries /// outputs that do not exist on Metal. Adding MPS dispatch to the CUDA op would require returning dummy tensors for every MPS call, which is confusing. The exact same design decision was made for the non-varlen path where  was created as a separate op. Does this reasoning hold for you, or would you prefer we use the existing CUDA op signature with -filled CUDA-specific outputs on MPS? Happy to change the approach if you have a preference.

**Comment by anagnorisis2peripeteia** [comment]
Updated in latest force-push (050be68):

1. **Metal 4.0 compatibility**: replaced `simdgroup_async_copy` double-buffered K prefetch with synchronous `BlockLoaderT` loading (`simdgroup_async_copy` was removed from `metal_simdgroup_matrix` in Metal 4.0 / macOS 26).

2. **Missing threadgroup barrier in fwd_sgmma**: K and V now share one smem buffer (single-buffer, no async prefetch). Added `threadgroup_barrier` at the end of the K-block loop to prevent the next iteration's K write from racing with the current iteration's V reads. Fixed 14 previously-failing multi-tile tests.

3. **Window backward masking bugs in `flash_attn_varlen_bwd_dkdv_sgmma`** (pre-existing in original commit, caught by the test suite):
   - Per-element mask had `wnd_left`/`wnd_right` conditions transposed.
   - Q-block skip conditions for `wnd_left`/`wnd_right` also had wrong directions.

All 85 `TestFlashAttentionVarlenMPS` tests pass.
